### ISSUE TITLE
SHAP values are not additive in a probability space

### ISSUE THREAD

**Issue by ireneisdoomed** [opentargets/issues#3755]
As a **developer**, I want to **normalize SHAP contributions to probability space while preserving additivity and directionality** because **users need to understand how each feature contributes to the final probability in an interpretable way**.

## Background

When using SHAP values to explain model predictions, the contributions are provided in log-odds space. However, users prefer to see contributions in probability space (e.g., "Why does this prediction have a probability of 0.6?"). Converting SHAP values directly to probabilities using the sigmoid function breaks the additivity property of SHAP values, meaning the contributions no longer sum to the final probability. This makes it difficult to interpret the results.

To address this, we propose a solution that:
1. **Preserves additivity**: Ensures that the contributions sum to the final probability.
2. **Preserves directionality**: Ensures that positive/negative SHAP values translate to positive/negative contributions in probability space.
3. **Distributes the base probability**: Allocates the base probability (inherent model bias) proportionally to the SHAP magnitudes, so the base is implicitly included in the contributions.

## Tasks

- [ ] Implement a `_normalise_feature_contributions` method to normalise SHAP contributions using the proposed solution.
- [ ] Update the `explain` method to pass the base log-odds to the normalisation function.
- [ ] Add unit tests to verify that contributions sum to the final probability and preserve directionality.

## Acceptance tests

How do we know the task is complete?

1. When I call the `explain` method, the output includes normalized probability contributions for each feature.
2. The sum of the normalized contributions matches the final probability (within a small tolerance, e.g., 1e-6).
3. Positive SHAP values result in positive contributions, and negative SHAP values result in negative contributions.
4. The base probability is distributed proportionally to the SHAP magnitudes, ensuring the contributions are interpretable.

**Comment by ireneisdoomed** [comment]
## Comparison between contributions for `2089b267ff0a27715af4b75d81abd834`

### Definitions
**Original logodds/probabilities**: Calculated following the approach in this [notebook](https://github.com/d0choa/l2g_widget_poc/blob/master/regenerate_shap.ipynb): we give the explainer a seed dataset based on randomly sampling the whole universe of locus to gene predictions.
- Probabilities are estimated by SHAP directly. Base value is not distributed across features

**Original logodds/probabilities**: the explainer uses data during training as a default seed. With this method, SHAP can't explain based in probabilities, so we approximate the probabilities, and split the base probability across features by following the method in this PR
- Probabilities are estimated by approximating the sigmoid function. Base value is distributed across features

### Comparison for top scoring gene (ENSG00000112514)

<img width="888" alt="Image" src="https://github.com/user-attachments/assets/4f0f3d24-973e-4994-9487-d3ee955c5b21" />

### Observations

- **_Marginal contributions change_**. For this locus, all 3 prioritised genes show different probabilities. I think this is largely due to the different approaches in estimating the contributions, more than the different approximation methods to get probabilities. To see this, look at the bar plot on the left, where I used log odds for both. `original` has the seed data, `new` doesn't.
- _**Distributing the base value can cause a change in the direction of the contribution**_. If you compare `proteinGeneCount` or `distanceSentinelFootprint` between the chart of log odds and the chart of probabiltiies for the new approach, you see that we can go from having a negative contribution to a positive one after sharing the base probability.
- **With the new approach, difference in contributions are less pronounced**. I think this is derived from the previous point. Negative contributions are less frequent, and therefore positive contributions are more even. For example, for the gene above:

<img width="252" alt="Image" src="https://github.com/user-attachments/assets/f568c17c-357b-4296-b272-c2a623b8c4f0" />

Based on this, my feeling is that it is important to understand what is the best strategy: if having a seed dataset or going with the default data.
Here's the df i sued to make the comparisons: 

[locus_comparison.csv](https://github.com/user-attachments/files/18688891/locus_comparison.csv)

<details close>
<summary>Code to reproduce</summary>
<br>

```python

import pandas as pd
import pyspark.sql.functions as f
import shap

from gentropy.common.session import Session
from gentropy.common.spark_helpers import pivot_df
from gentropy.dataset.l2g_feature_matrix import L2GFeatureMatrix
from gentropy.dataset.l2g_prediction import L2GPrediction
from gentropy.method.l2g.model import LocusToGeneModel
from gentropy.config import LocusToGeneConfig

session = Session()

features_list = LocusToGeneConfig(session).features_list
model = LocusToGeneModel.load_from_disk(
    "locus_to_gene_model/classifier.skops", features_list=features_list
)

pred = L2GPrediction.from_parquet(session, "l2g_predictions", model=model).filter(
    f.col("studyLocusId") == "2089b267ff0a27715af4b75d81abd834"
)
fm = L2GFeatureMatrix(
    _df=session.load_data("l2g_feature_matrix/"),
)
pred_feat = pred.add_features(fm)
explanations = pred_feat.explain()
exp = explanations.df.persist()

pivot_exp = exp.withColumn("feature", f.explode("features")).select(
    "studyLocusId",
    "geneId",
    "score",
    f.col("feature.name").alias("feature_name"),
    f.col("feature.scaledProbability").alias("new_proba"),
    f.col("feature.shapValue").alias("new_logodds"),
)

# original SHAP calculation
fm = fm.select("studyLocusId", "geneId", *features_list)
fm_pd = fm.limit(100_000).toPandas().drop(["studyLocusId", "geneId"], axis=1)
explainer_logodds = shap.TreeExplainer(model.model, data=fm_pd, model_output="raw")
explainer_proba = shap.TreeExplainer(
    model.model, data=fm_pd, model_output="probability"
)


subset = fm.filter(
    (f.col("studyLocusId") == "2089b267ff0a27715af4b75d81abd834")
).filter(
    f.col("geneId").isin(["ENSG00000112514", "ENSG00000112511", "ENSG00000197283"])
)
subset_pdf = subset.toPandas()
shap_values_logodds = explainer_logodds(
    subset_pdf.drop(["studyLocusId", "geneId"], axis=1)
)
shap_values_proba = explainer_proba(subset_pdf.drop(["studyLocusId", "geneId"], axis=1))

for i, feature in enumerate(features_list):
    subset_pdf[f"original_logodds_{feature}"] = [
        row[i].values for row in shap_values_logodds
    ]
    subset_pdf[f"original_proba_{feature}"] = [
        row[i].values for row in shap_values_proba
    ]
    subset_pdf.drop(feature, axis=1, inplace=True)
subset_pdf


# Convert from wide to long to compare values
def reshape_df_to_long(df, features_list):
    id_cols = ["studyLocusId", "geneId"]
    logodds_cols = [f"original_logodds_{feature}" for feature in features_list]
    proba_cols = [f"original_proba_{feature}" for feature in features_list]

    df_logodds = df[id_cols + logodds_cols].melt(
        id_vars=id_cols, var_name="feature_name", value_name="original_logodds"
    )
    df_logodds["feature_name"] = df_logodds["feature_name"].str.replace(
        "original_logodds_", ""
    )

    df_proba = df[id_cols + proba_cols].melt(
        id_vars=id_cols, var_name="feature_name", value_name="original_proba"
    )
    df_proba["feature_name"] = df_proba["feature_name"].str.replace(
        "original_proba_", ""
    )
    return pd.merge(df_logodds, df_proba, on=id_cols + ["feature_name"])


long_df = reshape_df_to_long(subset_pdf, features_list)


original_df = session.spark.createDataFrame(long_df)

combined_df = original_df.join(
    pivot_exp, on=["feature_name", "geneId", "studyLocusId"]
).select(
    "studyLocusId",
    "geneId",
    "score",
    "feature_name",
    "original_logodds",
    "new_logodds",
    "original_proba",
    "new_proba",
)

```
</details>

cc @addramir @d0choa

**Comment by ireneisdoomed** [comment]
## Background datasets

TLDR; Let's use the dataset used to train the model as the background dataset. This is logged to W&B every time we train the model. 

### What is it supposed to do?
It is basically a baseline for understanding how all features contribute to deviations from the model's average behaviour.
It sets a **reference** for SHAP to calculate the base value (the model's inherent bias) and to simulate "missing" feature values by replacing them with values from the background dataset to isolate the contribution of each feature.

### Should we use it?
**Yes.** In fact, this is required for most of SHAP's explainers, and the `TreeExplainer` we use is an exception.
Contrary to what I thought, _without a background dataset, SHAP relies on the model's internal structure_, not actual data, to estimate the base value. By internal structure I mean the graph of trees the model learns to make decisions.
These trees are obviously generated based on the training set, but they don't represent accurately the actual data distribution. 
They are less accurate because the model's internal structure doesn't know how often a certain feature values actually occurs. It also assumes that all features are uniformly distributed.
So providing a a background dataset ensures more accurate and meaningful explanations.

### Which one should we use then?
My understanding is that this dataset should represent how the actual data looks like, just like @d0choa discussed the other day. Different seed datasets will extract different contributions. In this [paper](https://arxiv.org/abs/2204.11351) they look at how the contributions fluctuate depending on the dataset we use as seed.

In the end, we need to have a representative sample of the training data. It doesn't need to be the entire dataset, but it should capture the overall distribution of features and classes. Because our training data is highly imbalanced, we should pass a similarly imbalanced dataset. I've done several experiments to see how the average bias changes with sampling:
| # Positives | # Negatives | Base value | Note                                    |
|-------------|-------------|------------|-----------------------------------------|
| 684         | 9316        | 0.07053582 | Full training set (downloaded from W&B) |
| 0           | 1,000       | 0.02048459 |                                         |
| 60          | 940         | 0.04878943 | Stratified random sampling           |
| 80          | 920         | 0.05495799 | Random sampling                         |
| 200         | 800         | 0.0900727  |                                         |
| 684         | 0           | 0.59068652 |                                         |

First row represents the entirety of my training set maintaing the real class imbalance. The other rows are tests where I can see that as the positive count grows, the base value also grows.

Based on this, and given that [performance is not an issue](https://github.com/shap/shap/issues/955), I would **use the full training set as the background dataset**. This ensures the base value and SHAP values are accurate as it represents the true Distribution.

**Comment by ireneisdoomed** [comment]
Based on the above comment and now that we've established we want to use data as seed, I wanted to **compare the methods to estimate probabilities from log odds**:
1. By using SHAP's native method (by setting the `model_output` parameter to `probability`)
2. By doing our own approximation (by extracting SHAP values as log odds and linearising with [this function](https://github.com/opentargets/gentropy/blob/7b9aa030aaada18d5afa71a6784276b89f3ea675/src/gentropy/dataset/l2g_prediction.py#L262))

![Image](https://github.com/user-attachments/assets/13ae1d45-257a-4325-b84d-9d9f48363810)

Probabilities are much more comparable and similar now that seed data is the same.
I've run this analysis for 100 random predictions and got a **correlation of 0.99 between both methods.**
My suggestion would be:
- I'm happy to see our estimation seems sensible, however for simplicity I would get probabilities from SHAP directly.
- I still think splitting the base value is a good idea for the sake of simplicity and user experience (users don't need to know about SHAP unless they are specifically interested).
